File: cockpit/api/cockpit_api/schemas/operator_access.py
"""Schemata für den Operator-Zugriff (ADR-0019)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from cockpit_api.services.operator_access import MIN_REASON_LENGTH


class OperatorAccessRequest(BaseModel):
    """Was der Betreiber angibt — **nicht** wer er ist.

    Das Feld `operator` stand hier und ist mit ADR-0020 D3 verschwunden: der
    Name reist signiert mit und landet im Audit des Kunden. Solange der
    Aufrufer ihn selbst setzt, ist die Auskunft „wer hat zugesehen" eine
    Behauptung. Er kommt jetzt aus der angemeldeten Sitzung.
    """

    #: Ein mitgeschicktes `operator` ist ein Fehler und kein stiller
    #: Nicht-Effekt: ein alter Aufrufer glaubte sonst, sein Name sei
    #: angekommen, und im Audit des Kunden stünde ein anderer.
    model_config = ConfigDict(extra="forbid")

    #: Die Mindestlänge steht auch im Dienst — dort ist sie die Autorität
    #: (sie muss für ein CLI gelten). Hier, damit die Oberfläche den Fehler
    #: bekommt, bevor der Zugriff ausgestellt wird.
    reason: str = Field(min_length=MIN_REASON_LENGTH, max_length=2000)
    ticket: str | None = Field(default=None, max_length=64)


class OperatorAccessOut(BaseModel):
    """Der Einlöseschein. Er wird **einmal** ausgeliefert und nicht gespeichert."""

    jti: str
    #: Die Assertion selbst. Sie steht in keiner Antwort ein zweites Mal: die
    #: Konsole speichert sie nicht, und sie ist nach sechzig Sekunden wertlos.
    assertion: str
    expires_at: datetime
    #: Die Adresse, die der Operator öffnet. Die Assertion steht im
    #: **Fragment** und nicht in der Abfrage: ein Fragment erreicht keinen
    #: Server-Log und keinen Referer-Header.
    redeem_url: str


class OperatorAccessGrantOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    jti: str
    operator: str
    reason: str
    ticket: str | None
    issued_at: datetime
    expires_at: datetime


__all__ = ["OperatorAccessGrantOut", "OperatorAccessOut", "OperatorAccessRequest"]

File: cockpit/web/src/api/operatorAccess.ts
import { get, post } from "./client";

export interface OperatorAccessGrant {
  jti: string;
  operator: string;
  reason: string;
  ticket: string | null;
  issued_at: string;
  expires_at: string;
}

export interface OperatorAccessOpened extends Pick<OperatorAccessGrant, "jti"> {
  /**
   * Der Einlöseschein. Er kommt **einmal** und wird nicht gespeichert — nach
   * sechzig Sekunden ist er wertlos.
   */
  assertion: string;
  expires_at: string;
  /** Die Adresse mit dem Schein im Fragment (nicht in der Abfrage). */
  redeem_url: string;
}

export interface OperatorAccessRequest {
  reason: string;
  ticket?: string | null;
}

export function openOperatorAccess(
  tenantId: string,
  body: OperatorAccessRequest,
): Promise<OperatorAccessOpened> {
  return post(`/api/tenants/${tenantId}/operator-access`, body);
}

export function listOperatorAccess(tenantId: string): Promise<OperatorAccessGrant[]> {
  return get(`/api/tenants/${tenantId}/operator-access`);
}
